### PR TITLE
[Fix] Appropriate whitespace missing in streaming output for Llama2, Mistral models

### PR DESCRIPTION
**Pull request by dbogunowicz**
Fix for: https://app.asana.com/0/1205229323407165/1205993428418769/f

## Testing

```python
from deepsparse import Pipeline
model_path = "zoo:opt-1.3b-opt_pretrain-pruned50_quantW8A8"
model_path = "zoo:llama2-7b-gsm8k_llama2_pretrain-pruned80_quantized"
pipeline = Pipeline.create(task="text-generation", model_path=model_path, sequence_length=64)
generations = pipeline(prompt="Hi, my name is Slim", streaming=True)
print(("").join([g.generations[0].text for g in generations]))
```

### Before:

```bash
.everyoneisamemberofthesamegroup,andthereare10membersinthegroup.
Ifyouareamemberofthegroup,andthereare10membersinthegroup,thenthegroupisdividedinto
```

### Now:

```bash
. everyone is a member of the same group, and there are 10 members in the group.
If you are a member of the group, and there are 10 members in the group, then the group is divided into
```

### Making sure that streaming for multiple prompts at once also works

```python
from deepsparse import Pipeline
prompts=["Hi, my name is Slim Shady", "Napoleon was"]
model_path = "zoo:llama2-7b-gsm8k_llama2_pretrain-pruned80_quantized"
pipeline = Pipeline.create(task="text-generation", model_path=model_path, sequence_length=64)

generations_first_prompt_only = list(pipeline(prompt=prompts[0], streaming=True))
generations_second_prompt_only = list(pipeline(prompt=prompts[1], streaming=True))
text_generated_first_prompt_only = "".join([g.generations[0].text for g in generations_first_prompt_only])
text_generated_second_prompt_only = "".join([g.generations[0].text for g in generations_second_prompt_only])
print(f"Text one: {text_generated_first_prompt_only}")
print(f"Text two: {text_generated_second_prompt_only}")

bag_of_words_first_prompt_only = [g.generations[0].text for g in generations_first_prompt_only]
bag_of_words_second_prompt_only = [g.generations[0].text for g in generations_second_prompt_only]

generations = pipeline(prompt=prompts, streaming=True)
bag_of_words = []
for r in generations:
    for gen in r.generations:
        text = gen.text
        bag_of_words.append(text)

assert sorted(bag_of_words_first_prompt_only+bag_of_words_second_prompt_only) == sorted(bag_of_words)
```

### CODE REVIEW

**Review by mgoin — APPROVED** [review]
Fixes my issue, thank you!

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 1]
```diff
@@ -601,6 +610,36 @@ def _stream_engine_outputs(
...
+            the list of last n tokens)
+        :param n: the number of last tokens to use for decoding
+        :return: the generated string
+        """
+        if not hasattr(self, "last_n_tokens"):
+            self.last_n_tokens = self.tokenizer(prompts).input_ids[-n:]
+        string_from_n_tokens = self.tokenizer.decode(
+            self.last_n_tokens, skip_special_tokens=True
+        )
+        self.last_n_tokens.append(generated_tokens[0])
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:636]: cannot update a pipeline level property for a specific inference run

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 2]
```diff
@@ -601,6 +610,36 @@ def _stream_engine_outputs(
...
+        string_from_n_plus_1_tokens = self.tokenizer.decode(
+            self.last_n_tokens, skip_special_tokens=True
+        )
+        self.last_n_tokens.pop(0)
+        return string_from_n_plus_1_tokens[len(string_from_n_tokens) :]
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:641]: slices are inneficient, is there something else we can do

**Review by bfineran — CHANGES_REQUESTED** [review]

File: src/deepsparse/transformers/pipelines/text_generation.py (hunk 2 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/pipelines/text_generation.py:644]: Should not be too much of an overhead (per: https://stackoverflow.com/questions/16891340/remove-a-prefix-from-a-string)

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 3]
```diff
@@ -582,8 +582,17 @@ def _stream_engine_outputs(
         for output in engine_outputs:
             generated_tokens, generated_logits, finished_reason = output
             logits = generated_logits if generation_config.output_scores else None
+            from transformers import LlamaTokenizer, LlamaTokenizerFast
```
[the remark this answers is not in the document]
Comment (dbogunowicz) [reply src/deepsparse/transformers/pipelines/text_generation.py:591]: @mgoin ?

[the remark this answers is not in the document]
Comment (mgoin) [reply src/deepsparse/transformers/pipelines/text_generation.py:591]: I saw it for other llama models, I think we just don't have a streaming test for them

**Review by mgoin — APPROVED** [review]

**Review by bfineran — APPROVED** [review]
LGTM - let's get this in for v2 in the next week

**Pull request merged**